File: LISTA1/batalhanaval.java
import java.util.Scanner;

public class BatalhaNaval {
    static final int NUM_JOGADORES = 2;
    static final int NUM_NAVIOS = 5;
    static final int TAMANHO_TABULEIRO = 5;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Bem-vindo ao jogo de Batalha Naval!!!");

        // Chamando tabuleiro
        char[][][] tabuleirosModelo = new char[NUM_JOGADORES][TAMANHO_TABULEIRO + 1][TAMANHO_TABULEIRO + 1];
        char[][][] tabuleirosJogo = new char[NUM_JOGADORES][TAMANHO_TABULEIRO + 1][TAMANHO_TABULEIRO + 1];
        for (int i = 0; i < NUM_JOGADORES; i++) {
            tabuleirosModelo[i] = gerarTabuleiro(TAMANHO_TABULEIRO);
            tabuleirosJogo[i] = gerarTabuleiro(TAMANHO_TABULEIRO);
        }

        // Coordenadas
        for (int jogador = 0; jogador < NUM_JOGADORES; jogador++) {
            System.out.println("\nJOGADOR " + (jogador + 1) + ", INSIRA AS COORDENADAS DO SEU NAVIO");

            for (int navio = 0; navio < NUM_NAVIOS; navio++) {
                boolean verificarNavio = false;
                while (!verificarNavio) {
                    System.out.print("Insira a localização do navio " + (navio + 1) + ": ");
                    verificarNavio = gerarNavio(tabuleirosModelo[jogador], scanner.nextInt(), scanner.nextInt());
                }
            }
            imprimirTabuleiro(tabuleirosModelo[jogador]);
        }

        // Inicializando jogo
        int[] contadorAcertos = new int[NUM_JOGADORES];
        int vencedor = -1, alvo = 0;
        while (vencedor == -1) {
            for (int jogador = 0; jogador < NUM_JOGADORES; jogador++) {

                if (jogador == 0)
                    alvo = 1;
                else if (jogador == 1)
                    alvo = 0;

                boolean verificarCoordenada = false;
                while (!verificarCoordenada) {
                    System.out.print("\nJogador " + (jogador + 1) + ", insira a linha/coluna do tiro: ");
                    verificarCoordenada = atirar(jogador, alvo, contadorAcertos, tabuleirosModelo, tabuleirosJogo, scanner.nextInt(), scanner.nextInt());
                }

                imprimirTabuleiro(tabuleirosJogo[alvo]);
                System.out.println("Navios restantes: " + (NUM_NAVIOS - contadorAcertos[jogador]));

                if (NUM_NAVIOS - contadorAcertos[jogador] == 0) {
                    vencedor = jogador;
                    break;
                }
            }
        }

        // Resultado
        System.out.println("\nJOGADOR " + (vencedor + 1) + " VENCEU! VOCÊ AFUNDOU TODOS OS NAVIOS DO OPONENTE!");
        System.out.println("Tabuleiros finais:\nJOGADOR 1:");
        imprimirTabuleiro(tabuleirosJogo[0]);
        System.out.println("\nJOGADOR 2:");
        imprimirTabuleiro(tabuleirosJogo[1]);

        scanner.close();
    	}

   //Geracao de tabuleiro
    public static char[][] gerarTabuleiro(int tamanho) {
        if (tamanho < 1 || tamanho > 9) {
            System.out.println("Erro: O tamanho deve estar entre 1 e 9. Retornando matriz vazia.");
            return null;
        }
        char[][] tabuleiro = new char[tamanho + 1][tamanho + 1];
        tabuleiro[0][0] = '*';

        for (int i = 1; i <= tamanho; i++) {
            tabuleiro[0][i] = (char) (i + '0');
            tabuleiro[i][0] = (char) (i + '0');
            for (int j = 1; j <= tamanho; j++)
                tabuleiro[i][j] = '-';
        }
        return tabuleiro;
    }

    //Imprimindo tabuleiro
    public static void imprimirTabuleiro(char[][] tabuleiro) {
        for (int i = 0; i < tabuleiro.length; i++) {
            for (int j = 0; j < tabuleiro[i].length; j++)
                System.out.print(tabuleiro[i][j] + " ");
            System.out.println();
        }
    }

    
    public static boolean gerarNavio(char[][] tabuleiro, int linha, int coluna) {
        if (linha < 1 || linha >= tabuleiro.length || coluna < 1 || coluna >= tabuleiro[0].length)
            System.out.println("Coordenadas inválidas. Escolha coordenadas diferentes.");

        else if (tabuleiro[linha][coluna] == '@')
            System.out.println("Você já posicionou um navio aqui. Escolha coordenadas diferentes.");

        else {
            tabuleiro[linha][coluna] = '@';
            return true;
        }
        return false;
    }

    //Atualizando tabuleiro de acordo com as entradas
    public static boolean atirar(int jogador, int alvo, int[] contadorAcertos, char[][][] tabuleirosModelo, char[][][] tabuleirosJogo, int linha, int coluna) {
        if (linha < 1 || linha >= tabuleirosModelo[0].length || coluna < 1 || coluna >= tabuleirosModelo[0][0].length)
            System.out.println("Coordenadas inválidas. Escolha coordenadas diferentes.");

        else {
            switch (tabuleirosModelo[alvo][linha][coluna]) {
                case 'X':
                case 'O':
                    System.out.println("Você já atirou neste local. Escolha coordenadas diferentes.");
                    break;

                case '-':
                    tabuleirosModelo[alvo][linha][coluna] = 'O';
                    tabuleirosJogo[alvo][linha][coluna] = 'O';
                    System.out.println("JOGADOR " + (++jogador) + " ERROU!");
                    return true;

                case '@':
                    tabuleirosModelo[alvo][linha][coluna] = 'X';
                    tabuleirosJogo[alvo][linha][coluna] = 'X';
                    contadorAcertos[jogador]++;
                    System.out.println("JOGADOR " + (++jogador) + " ACERTOU O NAVIO DO JOGADOR " + (++alvo) + "!!!");
                    return true;
            }
        }
        return false;
    }
}

File: LISTA2/batalhanaval.py
import random

NUM_JOGADORES = 2
NUM_NAVIOS = 5
TAMANHO_TABULEIRO = 5

# Funcao para gerar um tabuleiro
def gerarTabuleiro(tamanho):
    if tamanho < 1 or tamanho > 9:
        print("Erro: O tamanho deve estar entre 1 e 9. Retornando matriz vazia.")
        return None
    
    tabuleiro = [['-' for _ in range(tamanho + 1)] for _ in range(tamanho + 1)]
    tabuleiro[0][0] = '*'
    
    for i in range(1, tamanho + 1):
        tabuleiro[0][i] = str(i)
        tabuleiro[i][0] = str(i)
    
    return tabuleiro

# Funcao para imprimir um tabuleiro
def imprimirTabuleiro(tabuleiro):
    for linha in tabuleiro:
        print(' '.join(linha))

# Funcao para posicionar um navio no tabuleiro
def gerarNavio(tabuleiro, linha, coluna):
    if linha < 1 or linha >= len(tabuleiro) or coluna < 1 or coluna >= len(tabuleiro[0]):
        print("Coordenadas invalidas. Escolha coordenadas diferentes.")
    elif tabuleiro[linha][coluna] == '@':
        print("Voce ja posicionou um navio aqui. Escolha coordenadas diferentes.")
    else:
        tabuleiro[linha][coluna] = '@'
        return True
    return False

# Funcao para realizar um tiro
def atirar(jogador, alvo, contadorAcertos, tabuleirosModelo, tabuleirosJogo, linha, coluna):
    if linha < 1 or linha >= len(tabuleirosModelo[0]) or coluna < 1 or coluna >= len(tabuleirosModelo[0][0]):
        print("Coordenadas invalidas. Escolha coordenadas diferentes.")
    else:
        case = tabuleirosModelo[alvo][linha][coluna]
        if case in ['X', 'O']:
            print("Voce ja atirou neste local. Escolha coordenadas diferentes.")
        elif case == '-':
            tabuleirosModelo[alvo][linha][coluna] = 'O'
            tabuleirosJogo[alvo][linha][coluna] = 'O'
            print(f"JOGADOR {jogador + 1} ERROU!")
            return True
        elif case == '@':
            tabuleirosModelo[alvo][linha][coluna] = 'X'
            tabuleirosJogo[alvo][linha][coluna] = 'X'
            contadorAcertos[jogador] += 1
            print(f"JOGADOR {jogador + 1} ACERTOU O NAVIO DO JOGADOR {alvo + 1}!!!")
            return True
    return False

def main():
    print("Bem-vindo ao jogo de Batalha Naval!!!")
    
    # Chamando tabuleiro
    tabuleirosModelo = [[[None for _ in range(TAMANHO_TABULEIRO + 1)] for _ in range(TAMANHO_TABULEIRO + 1)] for _ in range(NUM_JOGADORES)]
    tabuleirosJogo = [[[None for _ in range(TAMANHO_TABULEIRO + 1)] for _ in range(TAMANHO_TABULEIRO + 1)] for _ in range(NUM_JOGADORES)]
    
    for jogador in range(NUM_JOGADORES):
        tabuleirosModelo[jogador] = gerarTabuleiro(TAMANHO_TABULEIRO)
        tabuleirosJogo[jogador] = gerarTabuleiro(TAMANHO_TABULEIRO)
    
    # Coordenadas
    for jogador in range(NUM_JOGADORES):
        print(f"\nJOGADOR {jogador + 1}, INSIRA AS COORDENADAS DO SEU NAVIO")
        
        for navio in range(NUM_NAVIOS):
            verificarNavio = False
            while not verificarNavio:
                linha = int(input(f"Insira a localizacao da linha do navio {navio + 1}: "))
                coluna = int(input(f"Insira a localizacao da coluna do navio {navio + 1}: "))
                verificarNavio = gerarNavio(tabuleirosModelo[jogador], linha, coluna)
        
        imprimirTabuleiro(tabuleirosModelo[jogador])
    
    # Inicializando jogo
    contadorAcertos = [0 for _ in range(NUM_JOGADORES)]
    vencedor = -1
    alvo = 0
    while vencedor == -1:
        for jogador in range(NUM_JOGADORES):
            if jogador == 0:
                alvo = 1
            elif jogador == 1:
                alvo = 0
            
            verificarCoordenada = False
            while not verificarCoordenada:
                print(f"\nJOGADOR {jogador + 1}, insira a linha do tiro: ")
                linha = int(input())
                print(f"Jogador {jogador + 1}, insira a coluna do tiro: ")
                coluna = int(input())
                verificarCoordenada = atirar(jogador, alvo, contadorAcertos, tabuleirosModelo, tabuleirosJogo, linha, coluna)
            
            imprimirTabuleiro(tabuleirosJogo[alvo])
            print(f"Navios restantes: {NUM_NAVIOS - contadorAcertos[jogador]}")
            
            if NUM_NAVIOS - contadorAcertos[jogador] == 0:
                vencedor = jogador
                break
    
    # Resultado
    print(f"\nJOGADOR {vencedor + 1} VENCEU! VOCE AFUNDOU TODOS OS NAVIOS DO OPONENTE!")
    print("Tabuleiros finais:\nJOGADOR 1:")
    imprimirTabuleiro(tabuleirosJogo[0])
    print("\nJOGADOR 2:")
    imprimirTabuleiro(tabuleirosJogo[1])

if __name__ == "__main__":
    main()
